### PR TITLE
Add experimental distributed tracing with OpenTelemetry

### PR DESCRIPTION
**Pull request by vfreex**
- pyartcd: Add tracing for build-sync pipeline job.
- doozer: Add tracing for releases:gen-payload command.

This can be enabled by setting env var `TELEMETRY_ENABLED=1`.

Inter-process trace context is propagated through env var `TRACEPARENT`.

Currently traces are printed to stderr. In the future production traces
should be forwarded to a collector.

Requires https://github.com/openshift-eng/art-tools/pull/99 to include pyartcd/elliott/doozer versions in traces.

### CODE REVIEW

**Comment by joepvd** [comment]
/approve

Looks like a great change to wet our feet in this direction. Gonna leave lgtm to others.

File: elliott/pyproject.toml [hunk 1]
```diff
@@ -47,6 +47,7 @@ dependencies = [
     "errata-tool ~= 1.31.0",
     "future",
     "koji >= 1.18",
+    "opentelemetry-sdk",
```
Comment (locriandev) [inline elliott/pyproject.toml:50]: Unused for the moment

File: doozer/doozerlib/cli/release_gen_payload.py [hunk 2]
```diff
@@ -380,24 +400,30 @@ async def generate_assembly_issues_report(self, assembly_inspector: AssemblyInsp
...
         """
         Mismatched siblings are built from the same repo but at a different commit
         """
-
+        span = trace.get_current_span()
         self.logger.debug("Checking for mismatched sibling sources...")
         group_images: List = assembly_inspector.get_group_release_images().values()
+        issues = []
         for mismatched, sibling in PayloadGenerator.find_mismatched_siblings(group_images):
-            self.assembly_issues.append(AssemblyIssue(
```
Comment (locriandev) [inline doozer/doozerlib/cli/release_gen_payload.py:393]: Is this change relevant for the sake of OpenTelemetry?

File: doozer/doozerlib/telemetry.py [hunk 3]
```diff
@@ -0,0 +1,94 @@
+import os
```
Comment (locriandev) [inline doozer/doozerlib/telemetry.py:1]: `telemetry.py` is basically copied across doozer/pyartcd, with some minor differences. Can this module be a common one, and placed in `art-common` when ready?

File: pyartcd/pyartcd/pipelines/build_sync.py [hunk 4]
```diff
@@ -392,10 +397,16 @@ async def build_sync(runtime: Runtime, version: str, assembly: str, publish: boo
...
+        "pyartcd.param.dry_run": runtime.dry_run,
+        "pyartcd.param.version": version,
+        "pyartcd.param.assembly": assembly,
+    })
     try:
         await pipeline.run()
         await pipeline.handle_success()
+        span.set_status(trace.StatusCode.OK)
 
     except (RuntimeError, ChildProcessError):
```
Comment (locriandev) [inline pyartcd/pyartcd/pipelines/build_sync.py:411]: Wouldn't it be worthy to `span.set_status(trace.StatusCode.ERROR)` here?

File: pyartcd/pyartcd/pipelines/build_sync.py [hunk 5]
```diff
@@ -392,10 +397,16 @@ async def build_sync(runtime: Runtime, version: str, assembly: str, publish: boo
...
-
+    span = trace.get_current_span()
+    span.set_attributes({
+        "pyartcd.param.dry_run": runtime.dry_run,
+        "pyartcd.param.version": version,
+        "pyartcd.param.assembly": assembly,
+    })
     try:
         await pipeline.run()
         await pipeline.handle_success()
```
Comment (locriandev) [inline pyartcd/pyartcd/pipelines/build_sync.py:408]: If this is going to replace Redis failure counting, we should also instrument `handle_failure()` to start tracking payload issues

**Comment by locriandev** [comment]
Great stuff! Thanks @vfreex. I left a few comments, but we're just starting to experiment so I'm also good with merging and adjusting later

File: doozer/doozerlib/cli/release_gen_payload.py (hunk 2 above)
Comment (vfreex) [reply doozer/doozerlib/cli/release_gen_payload.py:393]: This is an attempt to record assembly issues in span attributes. It is not actually consumed by anything, but I would like to keep this and see how easy or hard to filter those span on visualization systems (e.g. kibana) or monitoring systems.

File: doozer/doozerlib/telemetry.py (hunk 3 above)
Comment (vfreex) [reply doozer/doozerlib/telemetry.py:1]: Yes, I am planing to do that when art-common is ready.

File: pyartcd/pyartcd/pipelines/build_sync.py (hunk 4 above)
Comment (vfreex) [reply pyartcd/pyartcd/pipelines/build_sync.py:411]: This function is wrapped by a decorator, which will automatically set the result to trace.StatusCode.ERROR on exception.

File: pyartcd/pyartcd/pipelines/build_sync.py (hunk 5 above)
Comment (vfreex) [reply pyartcd/pyartcd/pipelines/build_sync.py:408]: Not at this moment. I would like to try that after when we have visualization and monitoring on OSD. Maybe a meter will be used for that.

File: doozer/doozerlib/cli/release_gen_payload.py [hunk 6]
```diff
@@ -32,6 +35,10 @@
 from doozerlib.util import find_manifest_list_sha
 
 
+TRACER = trace.get_tracer(__name__)
+METER = metrics.get_meter(__name__)
```
[the remark this answers is not in the document]
Comment (vfreex) [reply doozer/doozerlib/cli/release_gen_payload.py:39]: sure

File: elliott/pyproject.toml (hunk 1 above)
Comment (vfreex) [reply elliott/pyproject.toml:50]: ack

File: pyartcd/pyartcd/cli.py [hunk 7]
```diff
@@ -48,7 +50,9 @@ def print_version(ctx, param, value):
               help="[MULTIPLE] increase output verbosity")
 @click.pass_context
 def cli(ctx: click.Context, config: Optional[str], working_dir: Optional[str], dry_run: bool, verbosity: int):
-
+    # Initialize telemetry if needed
+    if os.environ.get("TELEMETRY_ENABLED") == "1":
```
[the remark this answers is not in the document]
Comment (vfreex) [reply pyartcd/pyartcd/cli.py:54]: sure, will add an option

File: doozer/doozerlib/cli/__init__.py [hunk 8]
```diff
@@ -105,6 +106,10 @@ def cli(ctx, **kwargs):
     global CTX_GLOBAL
     kwargs['global_opts'] = None  # can only be set in settings.yaml, add manually
 
+    # Initialize telemetry if needed
+    if os.environ.get("TELEMETRY_ENABLED") == "1":
```
[the remark this answers is not in the document]
Comment (vfreex) [reply doozer/doozerlib/cli/__init__.py:110]: sure, will add an option

**Pull request merged**